cassandra/io/twistedreactor.py

```diff
@@ -168,34 +169,46 @@ def _on_loop_timer(self):
 
 try:
     from twisted.internet import ssl
+    from OpenSSL import SSL
     import OpenSSL.crypto
     from OpenSSL.crypto import load_certificate, FILETYPE_PEM
 
     class _SSLContextFactory(ssl.ClientContextFactory):
-        def __init__(self, ssl_options, check_hostname, host):
+        def __init__(self, ssl_options, check_hostname, host, ssl_context=None):
             self.ssl_options = ssl_options
             self.check_hostname = check_hostname
             self.host = host
+            if ssl_context and ssl_context._verify_callback:
+                raise Exception(
+                    "'set_verify' should not be called on context for Twisted. Set verify mode using 'ssl_options'.")
+            self.ssl_context = copy.copy(ssl_context)
 
         def getContext(self):
-            # This version has to be OpenSSL.SSL.DESIRED_VERSION
-            # instead of ssl.DESIRED_VERSION as in other loops
-            self.method = self.ssl_options["ssl_version"]
-            context = ssl.ClientContextFactory.getContext(self)
-            if "certfile" in self.ssl_options:
-                context.use_certificate_file(self.ssl_options["certfile"])
-            if "keyfile" in self.ssl_options:
-                context.use_privatekey_file(self.ssl_options["keyfile"])
-            if "ca_certs" in self.ssl_options:
-                x509 = load_certificate(FILETYPE_PEM, open(self.ssl_options["ca_certs"]).read())
-                store = context.get_cert_store()
-                store.add_cert(x509)
-            if "cert_reqs" in self.ssl_options:
-                # This expects OpenSSL.SSL.VERIFY_NONE/OpenSSL.SSL.VERIFY_PEER
-                # or OpenSSL.SSL.VERIFY_FAIL_IF_NO_PEER_CERT
-                context.set_verify(self.ssl_options["cert_reqs"],
-                                   callback=self.verify_callback)
-            return context
+            if self.ssl_context:
+                if self.ssl_options and "cert_reqs" in self.ssl_options:
+                    cert_reqs = self.ssl_options["cert_reqs"]
+                else:
+                    cert_reqs = SSL.VERIFY_PEER | SSL.VERIFY_FAIL_IF_NO_PEER_CERT
+                self.ssl_context.set_verify(cert_reqs, callback=self.verify_callback)
```

Comment:
Passing a context like this is new behavior, so I was trying to pick a reasonable default for new behavior. I think that by default we should use strong security. If the default is to not call `set_verify`, then we'll by default allow an SSL connection to proceed if the certificate verification fails and the hostname verification fails. That's not strong security and I don't think most people would want that. Using strong security by default also means people won't normally have to pass extra options with `ssl options`. They can just set the context and be done with it.

Though you commenting on this line did make me realize we don't need the `| SSL.VERIFY_FAIL_IF_NO_PEER_CERT` part because it's ignored in client mode.

Reply:
Those options are documented here: https://www.openssl.org/docs/manmaster/man3/SSL_CTX_set_verify.html

The relevant documentation is:
SSL_VERIFY_NONE
Client mode: if not using an anonymous cipher (by default disabled), the server will send a certificate which will be checked. The result of the certificate verification process can be checked after the TLS/SSL handshake using the SSL_get_verify_result(3) function. The handshake will be continued regardless of the verification result.

SSL_VERIFY_PEER
Client mode: the server certificate is verified. If the verification process fails, the TLS/SSL handshake is immediately terminated with an alert message containing the reason for the verification failure. If no server certificate is sent, because an anonymous cipher is used, SSL_VERIFY_PEER is ignored.


As far as default behavior for other contexts / event loops, it depends on how the context is constructed. The old way of passing ssl options and we construct the context defaults to using no certificate verification. The new way of using a passed in context depends on what the user set on their context. The default SSLContext constructor defaults to no verification, but the documentation pretty clearly calls that out as being insecure and not recommended:
```When calling the SSLContext constructor directly, CERT_NONE is the default. Since it does not authenticate the other peer, it can be insecure, especially in client mode where most of time you would like to ensure the authenticity of the server you’re talking to. Therefore, when in client mode, it is highly recommended to use CERT_REQUIRED. ```
https://docs.python.org/3/library/ssl.html#verifying-certificates